design-proposals/2022-11-29-metal-port.md

```diff
@@ -0,0 +1,447 @@
+
+#### Required Changes
+
+This pass is the last version of the shaders and would consist of:
+- Modularization.  We can make it easier to reuse shaders by sharing code where useful.  Even putting the core of a given shader into a subroutine would make it easier to reuse.
+- Data driven styling.  How much of that logic can be pushed into the shaders?  The more work they do, the less the CPU does.
+- Optimization.  Apple provides tools for analyzing time spent in the shaders.  Some optimization here is always useful.
+- Packaging.  Normally Metal shaders are pre-compiled and shipped with the app binary.  We may want to support a pure source version.
+
+We'd also suggest writing a couple of examples here for implementing your own shaders.
+
+#### Benefits
+
+Optimization tends to get lost in these projects, but it yields dividends later on.  Less time spent on the CPU, in particular, makes room for more functionality.
+
+Adding your own shaders is one of the big goals.  If we can break up the common functionality in nice sub-routines that becomes much easier.  Ideally we can foster a community of incremental updates to add new features, rather than having to periodically do these big updates.
+
+### Bug Roundup
+
+Bugs in rendering SDKs can be very subtle.  The interaction between hardware, drivers, and software is complex and the patterns may not become clear until a certain number of devices are reached.  When tracking down obscure rendering bugs in WhirlyGlobe-Maply we will often have problems that can't be reproduced and have to guess at the cause from nearby crashes.  It's not as bad as it sounds.
+
+To address this, we suggest allocating a bug squishing pass a head of time.  This should be several months after the initial development is complete and at least one big deployment has occurred.  With WhirlyGlobe-Maply our clients ended up absorbing this pass, but it would be wiser to plan ahead.
+
+#### Required Changes
+
+This depends entirely on the bugs we find.
+
+#### Benefits
+
+Big development projects do generate bugs.  We can certainly depend on a good number of testers.  The community is active and involved.  This will help immensely.  But we do want the big companies to adopt this new version and we want to fix what they find.  As such, scheduling a post-development bug sprint with the team that does the work would be a good idea.
+
+## API Modifications
+
+At present we do not expect to change any existing APIs at the public level.  We do expect to add some new API methods and classes.  We would detail these as we proceeded with the implementation.
+
+## Migration Plan and Compatibility
+
+We do not intend for these changes to be incompatible with the existing version.
+
+## Rejected Alternatives
+
+The main alternative would be switching to an intermediate toolkit to implement both OpenGL and Metal.  The problem with that approach is toolkit size.  Quite a few users cannot abide with the toolkit growing considerably.
```

Comment:
OpenGL is the other renderer.

As for the bug squishing phase, I'm trying to address a pattern I've seen many times before.  To make it explicit, rather than implicit.

I don't disagree with your point on binary size.  In the abstract it shouldn't be a consideration.  But sometimes the practicalities have to dominate and this looks, to me, like one of those times.